File: tests/test_websocket_peer_e2e.js
// =============================================================================
// TD Engine — End-to-end WebSocket peer integration test
//
// Verifies that the C++ WebSocketPeer (src/net/websocket_peer.{h,cpp}) speaks
// the SAME wire format as the JS TDNet.RPC layer (web/net_websocket.js),
// which is locked by the 28 tests in tests/test_net_websocket.js.
//
// How it works:
//   1. Spawns a C++ test binary that hosts a WebSocketPeer on a random port.
//   2. The C++ binary prints "LISTENING <port>" once the listener is ready.
//   3. This Node script opens a real browser-style WebSocket to that port.
//   4. Sends a JSON-RPC request frame:  {"id":1,"m":"ping","a":[]}
//   5. The C++ RpcServer has a registered "ping" handler that replies with
//      the response frame:  {"id":1,"r":"pong"}
//   6. This script asserts it receives the exact response frame.
//   7. Sends a notify frame (no id):  {"m":"echo","a":["hello"]}
//   8. C++ handler broadcasts back to all peers; this script asserts it
//      receives the echoed notify.
//   9. Closes cleanly; the C++ binary exits 0.
//
// Build the C++ binary (from repo root):
//   g++ -std=c++17 -Wall -Wextra -O2 -Isrc \
//       tests/test_websocket_peer_e2e.cpp \
//       src/net/websocket_peer.cpp \
//       src/net/transport.cpp \
//       src/net/json_rpc.cpp \
//       tests/stub_logger.cpp \
//       -lpthread -o build/test_websocket_peer_e2e
//
// Run:
//   node tests/test_websocket_peer_e2e.js
// =============================================================================

'use strict';

const { spawn } = require('child_process');
const path = require('path');
const net = require('net');

const REPO_ROOT = path.resolve(__dirname, '..');
const CPP_BIN = path.join(REPO_ROOT, 'build', 'test_websocket_peer_e2e');

let pass = 0, fail = 0;
function assert(cond, msg) {
  if (cond) { pass++; console.log('PASS:', msg); }
  else      { fail++; console.error('FAIL:', msg); }
}

function findFreePort() {
  return new Promise((resolve, reject) => {
    const srv = net.createServer();
    srv.unref();
    srv.on('error', reject);
    srv.listen(0, '127.0.0.1', () => {
      const port = srv.address().port;
      srv.close(() => resolve(port));
    });
  });
}

async function main() {
  // 1. Find a free port (so the C++ binary can listen without collision).
  const port = await findFreePort();

  // 2. Spawn the C++ test server.
  const child = spawn(CPP_BIN, [String(port)], { stdio: ['pipe', 'pipe', 'pipe'] });
  let ready = false;
  let stderrBuf = '';

  child.stdout.on('data', (buf) => {
    const text = buf.toString();
    if (!ready && text.startsWith('LISTENING')) {
      ready = true;
      runClient(port).then(() => {
        // Tell the C++ server to exit cleanly.
        child.stdin.write('quit\n');
      }).catch((e) => {
        console.error('Client failed:', e);
        child.kill('SIGKILL');
        process.exit(1);
      });
    }
  });
  child.stderr.on('data', (buf) => { stderrBuf += buf.toString(); });
  child.on('exit', (code) => {
    if (stderrBuf.trim()) console.log('[C++ stderr]:', stderrBuf.trim());
    console.log(`[C++ exit] code=${code}`);
  });

  // Timeout watchdog — if the C++ binary doesn't start in 5s, abort.
  setTimeout(() => {
    if (!ready) {
      console.error('TIMEOUT: C++ server never said LISTENING');
      console.error('stderr:', stderrBuf);
      child.kill('SIGKILL');
      process.exit(1);
    }
  }, 5000);
}

async function runClient(port) {
  // Use Node's built-in WebSocket (Node 22+) or fall back to ws package.
  let WebSocketImpl;
  try {
    WebSocketImpl = require('ws');
  } catch {
    if (typeof WebSocket !== 'undefined') {
      WebSocketImpl = WebSocket;
    } else {
      throw new Error('No WebSocket implementation available (install "ws" or use Node 22+)');
    }
  }

  await new Promise((r) => setTimeout(r, 100));  // small grace period

  const ws = new WebSocketImpl(`ws://127.0.0.1:${port}/`);
  const received = [];

  await new Promise((resolve, reject) => {
    ws.on('open', () => {
      // Step 1: send a JSON-RPC request.
      ws.send(JSON.stringify({ id: 1, m: 'ping', a: [] }));
    });
    ws.on('message', (data, isBinary) => {
      const text = typeof data === 'string' ? data : data.toString();
      received.push(text);

      // Verify the response.
      if (received.length === 1) {
        let frame;
        try { frame = JSON.parse(received[0]); } catch (e) { reject(e); return; }
        assert(frame.id === 1 && frame.r === 'pong',
               `ping response: ${received[0]}`);

        // Step 2: send a notify (no id).
        ws.send(JSON.stringify({ m: 'echo', a: ['hello'] }));
      } else if (received.length === 2) {
        let frame;
        try { frame = JSON.parse(received[1]); } catch (e) { reject(e); return; }
        // The C++ handler echoes back as a notify with method "echoed".
        assert(frame.m === 'echoed' && Array.isArray(frame.a) && frame.a[0] === 'hello',
               `echo notify: ${received[1]}`);
        ws.close();
        resolve();
      }
    });
    ws.on('error', reject);
    setTimeout(() => reject(new Error('WebSocket timeout')), 4000);
  });

  // Step 3: summary.
  console.log('---');
  console.log(`WebSocket peer E2E: passed ${pass}, failed ${fail}`);
  if (fail > 0) process.exit(1);
}

main().catch((e) => { console.error(e); process.exit(1); });

File: tests/test_websocket_peer_e2e.cpp
// =============================================================================
// TD Engine — WebSocketPeer end-to-end test (C++ server side)
//
// Pairs with tests/test_websocket_peer_e2e.js. The JS script spawns this
// binary, opens a WebSocket to it, and verifies the JSON-RPC round-trip.
//
// Build (manual, for now — not in CMakeLists to keep CI build time short):
//   g++ -std=c++17 -Wall -Wextra -O2 -Isrc \
//       tests/test_websocket_peer_e2e.cpp \
//       src/net/websocket_peer.cpp \
//       src/net/transport.cpp \
//       src/net/json_rpc.cpp \
//       tests/stub_logger.cpp \
//       -lpthread -o build/test_websocket_peer_e2e
//
// Usage:
//   ./build/test_websocket_peer_e2e <port>
//
// Behavior:
//   1. Hosts a WebSocketPeer on <port>.
//   2. Registers a "ping" RPC handler that returns "pong".
//   3. Registers an "echo" notify handler that broadcasts "echoed" to all
//      peers with the original arg.
//   4. Prints "LISTENING <port>" to stdout so the JS client knows to connect.
//   5. Polls in a loop. Exits when stdin receives "quit" or after 10s.
// =============================================================================

#include "net/websocket_peer.h"
#include "net/transport.h"
#include "net/json_rpc.h"
#include "core/logger.h"

#include <atomic>
#include <chrono>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <string>
#include <thread>

#ifdef _WIN32
    #include <winsock2.h>
    #include <windows.h>
    #pragma comment(lib, "ws2_32.lib")
#else
    #include <unistd.h>
    #include <sys/select.h>
#endif

using td::WebSocketPeer;
using td::NetPacket;
using td::NetReliability;
using td::RpcServer;

int main(int argc, char** argv) {
    if (argc < 2) {
        fprintf(stderr, "usage: %s <port>\n", argv[0]);
        return 1;
    }
    int port = atoi(argv[1]);

    // Start the WebSocketPeer server.
    WebSocketPeer peer;
    if (!peer.host(port, 8)) {
        fprintf(stderr, "[server] host(%d) failed: %s\n",
                port, peer.lastError().c_str());
        return 1;
    }

    // Register RPC handlers.
    RpcServer::get().setPeer(&peer);

    RpcServer::get().registerMethod("ping",
        [](int peerId, const char* /*argsJson*/) {
            // Reply with a Response frame.
            RpcServer::get().sendResponse(1, "\"pong\"", peerId);
        });

    RpcServer::get().registerMethod("echo",
        [](int peerId, const char* argsJson) {
            // Echo back as a Notify to ALL peers (broadcast).
            // argsJson looks like `["hello"]` — we just pass it through.
            RpcServer::get().callRemote("echoed", argsJson, /*targetPeer=*/-1);
        });

    // Signal readiness so the JS test client can connect.
    printf("LISTENING %d\n", port);
    fflush(stdout);

    // Poll loop. Exit on stdin "quit" or after 10 seconds.
    auto start = std::chrono::steady_clock::now();
    for (;;) {
        // Check stdin for "quit".
#ifdef _WIN32
        // On Windows, use select() on stdin (handle 0).
        // Simpler: just poll with timeout, no stdin check.
        Sleep(10);
#else
        fd_set rfds;
        FD_ZERO(&rfds);
        FD_SET(STDIN_FILENO, &rfds);
        timeval tv = {0, 10000};  // 10ms
        if (select(STDIN_FILENO + 1, &rfds, nullptr, nullptr, &tv) > 0) {
            char buf[64];
            ssize_t n = read(STDIN_FILENO, buf, sizeof(buf) - 1);
            if (n > 0) {
                buf[n] = 0;
                if (strstr(buf, "quit")) break;
            }
        }
#endif

        peer.poll();
        RpcServer::get().update(
            (uint32_t)std::chrono::duration_cast<std::chrono::milliseconds>(
                std::chrono::steady_clock::now() - start).count());

        auto elapsed = std::chrono::steady_clock::now() - start;
        if (std::chrono::duration_cast<std::chrono::seconds>(elapsed).count() > 10) {
            fprintf(stderr, "[server] timeout\n");
            break;
        }
    }

    peer.disconnect();
    return 0;
}
